### PR TITLE
Load election from smartcard

### PR DESCRIPTION
**Pull request by benadida**

### CODE REVIEW

**Comment by beausmith** [comment]
@benadida - please fix failing tests before I review.

**Comment by benadida** [comment]
@beausmith fixed. That darn half-line that wasn't covered :)

File: src/App.tsx [hunk 1]
```diff
@@ -94,6 +96,42 @@ export class App extends React.Component<RouteComponentProps, State> {
         this.activateBallot(activationData)
       }
     }
+
+    if (cardData.t === 'admin') {
+      if (!this.state.election) {
+        if (longValueExists && !this.state.loadingElection) {
```
[the remark this answers is not in the document]
Comment (benadida) [reply src/App.tsx]: Discussed in person, the double if is to leave room for admin card after election is loaded.

File: src/App.tsx [hunk 2]
```diff
@@ -72,7 +74,7 @@ let checkCardInterval = 0
 export class App extends React.Component<RouteComponentProps, State> {
   public state: State = initialState
 
-  public processCardData = (cardData: CardData) => {
+  public processCardData = (cardData: CardData, longValueExists: boolean) => {
```
[the remark this answers is not in the document]
Comment (benadida) [reply src/App.tsx]: 👍

File: src/App.tsx [hunk 3]
```diff
@@ -124,20 +162,7 @@ export class App extends React.Component<RouteComponentProps, State> {
...
-            const cardData = JSON.parse(resultJSON.shortValue) as CardData
-            this.processCardData(cardData)
-          }
-        })
-        .catch(() => {
-          // if it's an error, aggressively assume there's no backend and stop hammering
-          window.clearInterval(checkCardInterval)
-        })
-    }, 1000)
+    this.startPolling()
```
[the remark this answers is not in the document]
Comment (benadida) [reply src/App.tsx]: Discussed in person, the stopping is only when the backend is absent.

File: src/AppAdminCard.test.tsx~ [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+it(`Admin card fetches election`, async (done) => {
+  const div = document.createElement('div')
+  // @ts-ignore - App expects ReactRouter props, but are unnecessary for this test.
+  const app = (ReactDOM.render(<App />, div) as unknown) as App
+
+  app.fetchElection = jest.fn().mockResolvedValue(election)
+  app.setElection = jest.fn()
+  
+})
```
[the remark this answers is not in the document]
Comment (benadida) [reply src/AppAdminCard.test.tsx~]: correct, I need to remove that stupid backup file.

File: src/AppCard.test.tsx [hunk 5]
```diff
@@ -59,34 +59,62 @@ it(`CardData processing processes card data properly`, () => {
...
 })
+
+it(`Calls fetch on fetchElection`, () => {
+  fetchMock.resetMocks()
+
+  fetchMock.mockResponse(JSON.stringify(election))
+
+  const div = document.createElement('div')
+  // @ts-ignore - App expects ReactRouter props, but are unnecessary for this test.
+  const app = (ReactDOM.render(<App />, div) as unknown) as App
```
[the remark this answers is not in the document]
Comment (benadida) [reply src/AppCard.test.tsx]: Yeah, sadness, let's figure it out later.

**Comment by benadida** [comment]
> Looks like you decided not to set default for `longValueExists `, but I'm happy with the object argument for clarity!

yeah default was only useful in tests, so that actually felt counter-productive. I wanted the tests to be more explicit.

**Pull request merged**